pageserver/src/tenant/timeline.rs

```diff
@@ -2651,6 +2662,30 @@ impl Timeline {
         let layers = self.layers.read().unwrap();
 
         let mut max_deltas = 0;
+        let wanted_image_layers = self.wanted_image_layers.lock().unwrap();
+        if let Some((cutoff_lsn, wanted)) = &*wanted_image_layers {
+            let img_range =
+                partition.ranges.first().unwrap().start..partition.ranges.last().unwrap().end;
+            if wanted.overlaps(&img_range) {
+                //
+                // gc_timeline only pays attention to image layers that are older than the GC cutoff,
+                // but create_image_layers creates image layers at last-record-lsn.
+                // So it's possible that gc_timeline decides that it wants new image layer to be created for a key range,
+                // and on next compaction create_image_layers creates the image layer.
+                // But on next GC cycle, gc_timeline still wants the new image layer to be created,
+                // because the newly created image layer doesn't help to remove the delta layer,
+                // until the newly created image layer falls off the PITR horizon.
```

Comment:
Sorry, do not have some good idea how to implement such test.
There is not way now to force Neon ti create some particular layers. It depends on a lot of factor and is not deterministic.
So it is not clear how to model such scenario when mage layer should not be generated because we already have more recent image layer which not yet pass GC PiTR boundary.

What I can do is just to insert in create_image_layers some verification code which checks that no newer image layer exists. But it will actually duplicate just inserted check. So it is not clear what can be achieve with this test.

So we have random test_gc_old_layers.py test which just do overlapped insertions and check total database size.
Once test is completed I can add some check which will traverse layers on the disk and check that there are no two image layers with the same key range not interleaved by any delta layers. But it is actually right way to introduce yet another fluky test. And there i absolutely no warranty that such non-deterministic test can actually reproduce this situation.

Reply:
I managed to create test reproducing the problem: it is similar with test_gc_old_layers with two differences:
1. gc&compaction are performed twice
2. PiTR interval is extended from 1 to 10 seconds
Also I have to add in  image_layer.rs warning if existed image layer is overwritten.
I wonder if overwriting existed layers is possible without  forcing generation of wanted image layers?
In case of pageserver abnormal restart we should remove all layer files beyand disk_consistent_lsn.
Can in happen in case of remote layers downloaded from S3?

In any case - this test doesn't pass without this added check that image layer with lsn>=cutoff_lsn exists.
And it is passed with this check added.

Reply:
Perhaps run test_gc_old_layers.py for longer, and with non-zero pitr interval?

Reply:
It is done using test_gc_feedback test